File: app/src/main/java/com/harnet/devnews/service/ParseService.java
package com.harnet.devnews.service;

import org.json.JSONArray;
import org.json.JSONException;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutionException;

public class ParseService {
    private WebContentDownloader webContentDownloader = new WebContentDownloader();

    //get list of articles URL
    public List<URL> getArticlesURLs(String url){
        List<URL> articlesURLs = new ArrayList<>();

        // get content from a page
        String content = null;
        try {
            content =  new WebContentDownloader().execute(url).get();
        } catch (InterruptedException | ExecutionException e) {
            e.printStackTrace();
        }

        // get articles id
        if(content != null){
            try {
                JSONArray jsonArray = new JSONArray(content);
                for (int i=0; i< jsonArray.length(); i++){
                    int articleId = Integer.parseInt(jsonArray.getString(i));
                    // TODO think about do it final or move to ENUM
                    URL atricleURL = new URL("https://hacker-news.firebaseio.com/v0/item/"+ articleId + ".json?print=pretty");
                    articlesURLs.add(atricleURL);
                }
            } catch (JSONException | MalformedURLException e) {
                e.printStackTrace();
            }
        }else{
            throw new RuntimeException("Can't get a content");
        }

        return articlesURLs;
    }

    // get an article details
    public String parse(String urlString){
        StringBuilder site = new StringBuilder();
        try {
            URL url = new URL(urlString);
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            InputStream in = connection.getInputStream();
            InputStreamReader reader = new InputStreamReader(in);
            int data = reader.read();

            while (data != -1) {
                char current = (char) data;
                site.append(current);
                data = reader.read();
            }
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
        return site.toString();
    }
}

File: app/src/main/java/com/harnet/devnews/service/ParseService.kt
package com.harnet.devnews.service

import android.util.Log
import org.json.JSONArray
import org.json.JSONException
import java.io.IOException
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL
import java.util.*
import java.util.concurrent.ExecutionException

class ParseService {
    //TODO implement argument with max parsed articles list

    // get articles ids to Retrofit
    fun getArticlesIDs(url: String?, articlesToShow: Int): MutableList<String>? {
        val articlesIDs = mutableListOf<String>()
        // get content from a page
        var content: String? = null
        try {
            content = WebContentDownloader().execute(url).get()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        } catch (e: ExecutionException) {
            e.printStackTrace()
        }

        try {
            val jsonArray = JSONArray(content)
            for (i in 0 until articlesToShow) {
                articlesIDs.add(jsonArray.getString(i))
            }
        } catch (e: InterruptedException) {
            e.printStackTrace()
        } catch (e: ExecutionException) {
            e.printStackTrace()
        }
        return articlesIDs
    }

    //get list of articles URL for parser
    fun getArticlesURLs(url: String?, articlesToShow: Int): MutableList<URL>? {
        // get content from a page
        var content: String? = null
        try {
            content = WebContentDownloader().execute(url).get()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        } catch (e: ExecutionException) {
            e.printStackTrace()
        }
        return content?.let { getAticlesIds(it, articlesToShow) }
    }

    // get articles ids for parser
    private fun getAticlesIds (content: String, articlesToShow: Int): MutableList<URL> {
        val articlesURLs: MutableList<URL> = ArrayList()

        try {
            val jsonArray = JSONArray(content)
            for (i in 0 until articlesToShow) {
                val articleId = jsonArray.getString(i).toInt()
                // TODO think about do it final or move to ENUM
                val atricleURL = URL("https://hacker-news.firebaseio.com/v0/item/$articleId.json?print=pretty")
                articlesURLs.add(atricleURL)
            }
        } catch (e: JSONException) {
            e.printStackTrace()
        } catch (e: MalformedURLException) {
            e.printStackTrace()
        }

        return articlesURLs
    }

    // parse site content
    fun parse(urlString: String?): String {
        val site = StringBuilder()

        try {
            val url = URL(urlString)
            val connection = url.openConnection() as HttpURLConnection
            val iN = connection.inputStream
            val reader = InputStreamReader(iN)
            var data = reader.read()

            while (data != -1) {
                val current = data.toChar()
                site.append(current)
                data = reader.read()
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return site.toString()
    }

    fun parseImages(pageContent: String?): List<String>? {
        var imagesURL = mutableListOf<String>()
//        println(pageContent)
        val testLink3 = pageContent as CharSequence
        val word = "jpg"
//        val matcher = "(?i)(?<!\\p{L})$word(?!\\p{L})".toRegex()
        val matcher = "http(.*?)jpg".toRegex()
        val articleImage = matcher.findAll(testLink3).map { it.value }.toList()
//        Log.i("JPGimages", "parseImages: "  + matcher.findAll(testLink3).map { it.value }.toList())
//        Log.i("JPGimages", "parseImages: counter" + matcher.findAll(testLink3).count() )
            // => [cBa, Cba, cbA]
            // => 3
        imagesURL.add(articleImage[0])

        return imagesURL
    }
}